budgetkey_data_pipelines/pipelines/budget/national/changes/processed/extract-change-groups.py

```diff
@@ -0,0 +1,288 @@
+                if c['budget_code'].startswith('0047') and c['leading_item'] != 47
+                if sum(c[field] * c[field] for field in value_fields) > 0
+            ]
+            logging.debug('reserve date: kind:%s num:%s' % (
+                date_kind, len(date_reserve)
+            ))
+            num_found = 0
+            i = 0
+            for comb_size in range(2, min(len(date_reserve) + 1, 7)):
+                done = False
+                while not done:
+                    not_selected = list(
+                        x for x in date_reserve
+                        if x['trcode'] not in selected_transfer_codes
+                    )
+                    logging.debug('len(not_selected)=%d' % len(not_selected))
+                    date_groups = combinations(not_selected, comb_size)
+                    found = None
+                    done = True
+                    try:
+                        while True:
+                            i += 1
+                            if i % 100000 == 0:
+                                logging.debug('%s %s %s %s' % (
+                                    date, len(date_reserve), num_found, i
+                                ))
+                            group = date_groups.send(found)
+                            found = False
+                            sumvec = sum(
+                                sum(c.get(x, 0) for c in group) ** 2
+                                for x in value_fields
+                            )
+                            if sumvec == 0:
+                                pending = group[0]['pending']
+                                transfer_codes = set(x['trcode'] for x in group)
+                                if len(selected_transfer_codes & transfer_codes) > 0:
+                                    continue
+                                selected_transfer_codes.update(transfer_codes)
+                                transfer_codes = list(transfer_codes)
+                                transfer_codes.sort()
+                                num_found += len(transfer_codes)
+                                to_append = {
+                                    'transfer_ids': transfer_codes,
+                                    'date': date,
+                                    'pending': pending,
+                                }
+                                result.append(to_append)
+                                found = True
+                    except StopIteration:
+                        pass
+        for change in changes:
+            if not change['trcode'] in selected_transfer_codes:
+                selected_transfer_codes.add(change['trcode'])
+                result.append({
+                    'transfer_ids': [change['trcode']],
+                    'date': change['date'],
+                    'pending': change['pending']
+                })
+        return result
+
+    def build_groups(groups, changes):
```

Comment:
most of this code is not necessary, as we need in the output just these columns:
`leading_item, request_code, transaction_id, date, pending`

So for each request we'd have its transaction id.